### PR TITLE
The Genesis of Generic Germination

### PR DESCRIPTION
**Pull request by varkor**
*Long had its coming been foretold: a collaborative effort with @yodaldevoid, set in motion by @jplatte, to beget a new Kind: one of a very different Sort to those that come before it. Amidst promises of ineffable powers previously thought unobtainable, few dared believe that the prophecies were true. But as they gazed upon that which claimed to be the Beginning, a few gentle sparks of hope fluttered deep within. It was not Time yet. But it was a Sign. And maybe, for some, that was enough.*

There's a long way to go, but we're at the point where we would benefit from GitHub's reviewing capabilities.

r? @eddyb

### CODE REVIEW

**Comment by kennytm** [comment]
(Somehow I read the title as "Genesis of Germanic Germination" and thought why we're adding i18n into rustc itself 😐)

File: src/librustc/hir/mod.rs
Comment (petrochenkov) [inline src/librustc/hir/mod.rs]: The expression span is available through `AnonConst`, is it necessary to duplicate it here?
(If not, then `struct ConstArg` can be refactored away as well.)

Comment (petrochenkov) [reply src/librustc/hir/mod.rs]: Ah, I see, it's available, but not without side tables.

File: src/libsyntax/parse/parser.rs
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Not sure why `param` exists, it should always be `true`.
Also, `< const` is enough to identify a generic parameter, no need for look_ahead(2).

File: src/libsyntax/parse/parser.rs
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Leading attributes before generic parameters are allowed.
This case means something like `<#[attr]>`, i.e. no parameters at all.
I'd keep the old wording (maybe with replacing "lifetime parameters" with "generic parameters").

File: src/libsyntax/parse/parser.rs
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Parameters are kept in a single list now, so the order can be enforced in one place in AST validation.

File: src/libsyntax/parse/parser.rs
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: If `S<MY_CONST>` is supposed to work, then it can use same approach as identifiers in patterns.
In AST it's `PatKind::Ident`, but in HIR it's turned into a proper pattern based on resolution, for const arguments it would be something like `ast::GenericArg::Ident`.
I agree this is not necessary to do in this PR though.

File: src/libsyntax/parse/parser.rs
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: This will hit unreachable on `S<-1>`.
I'd just use `parse_expr` if `self.check_const_arg()` returns true.

(Unless you still have plans regarding turbofish, in this case a comment here explaining "why not `parse_expr`" would be appropriate.)

File: src/libsyntax/parse/token.rs
Comment (petrochenkov) [inline src/libsyntax/parse/token.rs]: There are couple more non-controversial additions to this list - `NtExpr`, `NtBlock`.

**Comment by varkor** [comment]
(Note for triage: I'm going to get around to addressing these comments and hopefully fixing the existing crashes soon.)

File: src/libsyntax/parse/parser.rs
Comment (varkor) [reply src/libsyntax/parse/parser.rs]: > Not sure why `param` exists, it should always be `true`.

I have no idea now either; removing.

File: src/libsyntax/parse/parser.rs
Comment (varkor) [reply src/libsyntax/parse/parser.rs]: Ah, this was an `unreachable` I missed when filling things out. `parse_expr` should do fine for now.

**Comment by TimNN** [comment]
Ping from triage! Since it's been I while, I was wondering if you could give us a small update on the status of this PR?

**Comment by varkor** [comment]
I'm going to start investigating the ICE again soon (been busier than I expected). I think @eddyb's going to be leaving some comments soon as well.

**Comment by TimNN** [comment]
Ping from triage! Since it's been I while, I was wondering if you could give us a small update on the status of this PR?

**Comment by varkor** [comment]
I've made progress in my local branch since last time, though not quite ready to push. I think a full review will be unlikely before the 2018 edition though, as things are getting busy close to the deadline.

**Comment by TimNN** [comment]
@vakor: Thanks for the update!

**Comment by andreytkachenko** [comment]
Will this PR solve the #43408 issue?
If yes then may be take a subset from the PR for solving that issue and then bring whole when it will be done?

**Comment by varkor** [comment]
@andreytkachenko: yes, this will allow arrays to be generic in their length parameters. However, the implementation all very much depends on the PR as a whole (the independent parts are the simpler parts), so there's no way to split it out.

We should be able to start making progress on this again after the 2018 edition is released.

**Comment by andreytkachenko** [comment]
Thank you, @varkor, for explanation

**Comment by eddyb** [comment]
@andreytkachenko @varkor Sorry for the bad news, but #43408 is about array lengths depending on generics *at all* (e.g. `[u8; size_of::<T>()]` today).
This PR will allow writing `[T; N]` where `N` is a `const` parameter, but *it won't work*.

That's because lazy normalization is required so we can get generic parameters anywhere near array length expressions without creating dependency cycles.
And only @nikomatsakis can tell you when lazy normalization will happen.

File: src/librustc/hir/lowering.rs
Comment (eddyb) [inline src/librustc/hir/lowering.rs]: This is surprising - these constants/statics should *not* be imported directly.

File: src/librustc_typeck/astconv.rs
Comment (eddyb) [inline src/librustc_typeck/astconv.rs]: How would this work? The `def_id` is that of the `AnonConst`, it can't be a parameter.
Maybe you want instead to check whether the expression *inside* `AnonConst` is an `ExprKind::Path` that has the `Def::ConstParam` resolution?

File: src/librustdoc/clean/mod.rs
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: Does this need special work? I would expect `fmt_const_val` below to "just work".

File: src/libserialize/serialize.rs
Comment (eddyb) [inline src/libserialize/serialize.rs]: Huh, I wouldn't think `PhantomData` would be relevant.
What about just fully implementing encoding/decoding for PhantomData, regardless of parameter?

**Review by eddyb — CHANGES_REQUESTED** [review]

**Comment by eddyb** [comment]
Other than the comments I left, I think this is good to go, after a rebase!

r? @nikomatsakis (at least a quick look would be greatly appreciated)

File: src/librustc/ty/relate.rs
Comment (oli-obk) [reply src/librustc/ty/relate.rs]: Values of layout `Scalar` are guaranteed `ConstValue::Scalar` for constants. Only statics don't normalize.

If other kinds of values can't be used in const generics, we should indeed bug out for them for now, there's no way for an int, float, bool or char constant to be anything but `ConstValue::Scalar(Scalar::Bits { .. })`. The type system will even guarantee that the `size` field of `Bits` must be equal if the types are equal, so you can add an assert for that, too.

Comment (eddyb) [reply src/librustc/ty/relate.rs]: Ah, so we could be limiting the type elsewhere, interesting.

File: src/librustdoc/clean/mod.rs
Comment (varkor) [reply src/librustdoc/clean/mod.rs]: Ah, that's true.

File: src/librustc/ty/subst.rs
Comment (varkor) [reply src/librustc/ty/subst.rs]: I don't think so. I've added a stub for the method for now.

**Comment by Mark-Simulacrum** [comment]
@varkor Will you get a chance to look at this shortly? It looks like there's a merge conflict and a CI failure from 5 days ago.

**Comment by varkor** [comment]
@Mark-Simulacrum: hopefully this weekend, but there are still some remaining issues I have to sort out before it's ready for a second review yet.

**Comment by nikomatsakis** [comment]
@varkor (ps, sorry for being slow, excited to review this though! Sounds like maybe I should wait though until you resolve a few things?)

**Comment by Aaronepower** [comment]
Triage; @varkor Have you been able to get back to this PR?

**Comment by varkor** [comment]
Sorry, I am going to get to this. I've been quite busy recently. This is my priority when I get time.

File: src/librustc/infer/combine.rs
Comment (nikomatsakis) [reply src/librustc/infer/combine.rs]: Answer: the type relation stuff tracks variance, and the folder does not. I've fixed this multiple times, but always in branches I wound up abandoning. Should just do it on a standalone branch at some point.

**Review by nikomatsakis — APPROVED** [review]
Did a first round of review. Looks excellent. Going to build locally also to investigate.

**Comment by felix91gr** [comment]
Thanks to all involved for your work on this. Remember to take a break from time to time, and I wish you have a good time in these times of celebration 😊 💐

File: src/librustc_typeck/collect.rs
Comment (varkor) [inline src/librustc_typeck/collect.rs]: Did the `bug` actually get triggered with invalid args?

File: src/librustc_typeck/astconv.rs
Comment (varkor) [inline src/librustc_typeck/astconv.rs]: This will change the order of errors, which might be misleading as they'll no longer appear in the same order they are written. Probably not an issue, but worth noting in a comment at least.

File: src/librustc_mir/monomorphize/item.rs
Comment (varkor) [inline src/librustc_mir/monomorphize/item.rs]: These `unwrap`s make me uncomfortable.

File: src/librustc_mir/monomorphize/item.rs
Comment (varkor) [inline src/librustc_mir/monomorphize/item.rs]: *`": "`

File: src/librustc/ty/fold.rs
Comment (varkor) [inline src/librustc/ty/fold.rs]: Ah, good catch! I had completely missed this, which would explain the errors.

File: src/librustc_typeck/collect.rs
Comment (yodaldevoid) [reply src/librustc_typeck/collect.rs]: Yes, it was getting triggered even after printing the errors created in the previous commit.

**Comment by varkor** [comment]
An brief update on progress for those who are following along:

@yodaldevoid fixed some of the remaining major crashes and now some of the const generic tests are compiling and passing 🎉

We've been working on fixing the remaining failing tests and hopefully soon we'll see the PR passing on Travis. After that, we're going to need to add more tests (there are only 6 const generic tests right now) to increase the coverage. I think then we'll be ready for another review. We can see the light at the end of the tunnel now!

**Comment by mark-i-m** [comment]
Thanks for all the hard work @varkor and @yodaldevoid! I’m sure I speak on behalf of many who deeply appreciate it :)

**Comment by felix91gr** [comment]
> Thanks for all the hard work @varkor and @yodaldevoid! I’m sure I speak on behalf of many who deeply appreciate it :)

You sure do, you beat me to it!

Thanks to everyone involved in this, specially @varkor and @yodaldevoid <3

I will use this for fun and learning so much that I can't quite believe that it will actually exist. Thanks, truly ☺️

File: src/librustc_typeck/diagnostics.rs
Comment (goffrie) [inline src/librustc_typeck/diagnostics.rs]: typo: `[[u8; WIDTH]; HEIGHT]`?

**Comment by Dylan-DPC** [comment]
ping from triage @varkor any updates on this? also conflicts are piling up :fearful:

**Comment by petrochenkov** [comment]
@varkor @yodaldevoid 
What's the point in keeping an "epic" PR open for months? <sub><sup>(Beside collecting likes.)</sup></sub>
Large part of this work could already land if it was done piecemeal + reviewing would be easier and more reliable.
For example, 1) Parser/AST + error during lowering to HIR -> 2) Lowering to HIR + error during astconv -> 3) Astconv and all the later utility passes with unification parts stubbed out somehow -> 4) Everything else.

**Comment by yodaldevoid** [comment]
You have finally caught on, this was all a ploy to get more likes!

All joking aside, this PR has run on far too long in some part from my lack of time to work on it for months, varkor's lack of time to work on it for months, and a little bit of my own perfectionism (I cannot speak to varkor in this regard). As you rightly point out, it is high time to get this specific PR through. I will rebase on master in the next day or so and then we can see about possibly splitting up the changes.

**Comment by yodaldevoid** [comment]
Ok, that may have been the most annoying rebase I have had to deal with so far, but it is done.

We have a few failing tests outside of the const-generic ones. Most of these tests have to do with catching problems with the order of generics as that has been messed with.

**Comment by lifeforfun** [comment]
Please don't give Rust too much sytactic suger and symbles, I don't have enough hair to lean it. Shouldn't a programming language stay simple and give developer to take care of their business?

**Comment by crlf0710** [comment]
Tip for people who is watching the progress of this PR, or const generics in general:

This epic PR is being split into parts. #58191 is its first part.

**Comment by leonardo-m** [comment]
How many/large parts are left?

**Comment by bemeurer** [comment]
#58503 is merged! GETTING THERE!

**Comment by mark-i-m** [comment]
Request: could you make a new post instead of updating the old one? Github doesn't send emails for updates. Thanks!

**Comment by varkor** [comment]
To summarise, this PR is being split out into smaller ones.
- (Merged) #58191: AST changes.
- (Merged) #58503: HIR changes.
- (Pending) #58581: HIR follow-up.
- (Pending) #58583: `ty` changes.

After this I plan to at least pull these into separate PRs:
- Inference changes.
- rustdoc changes.
- Addressing the remaining `FIXME`s.

We'll then be in a position to start implementing more tests and fixing the bugs in the implementation that haven't been spotted yet. My preference is to keep this PR open until the changes contained in it have been merged (separately), as it makes for a useful tracking PR for those interested in the implementation.

**Comment by mark-i-m** [comment]
#58583 merged :tada:

**Comment by varkor** [comment]
https://github.com/rust-lang/rust/pull/59008 has been opened with the changes to type inference.

**Comment by varkor** [comment]
#59008 is blocked on some refactoring to consts for now, so https://github.com/rust-lang/rust/pull/59170 has been opened in the meantime extending the support for const generics in rustdoc.

**Comment by Dylan-DPC** [comment]
ping from triage @varkor i'm assuming this is still blocked on #59170 ?

**Comment by varkor** [comment]
This is currently blocked on https://github.com/rust-lang/rust/pull/59008 (https://github.com/rust-lang/rust/pull/59170 was merged), which in turn is currently blocked on https://github.com/rust-lang-nursery/ena/pull/22 (but which is expected to be merged soon).

**Comment by varkor** [comment]
This is partially for my own reference, but as far as I see it the remaining steps after https://github.com/rust-lang/rust/pull/59008 are:
- Resolve the existing `FIXME(const_generics)` issues in the code.
- Make sure that const parameters in patterns are handled correctly.
- Add an error for const parameters depending on type parameters.
- Add (many) more tests.

That is, https://github.com/rust-lang/rust/pull/59008 contains most of the remaining parts of this pull request.

**Comment by mark-i-m** [comment]
#59008 has merged :tada:

**Comment by mark-i-m** [comment]
IIUC, some refactoring work and something related to canonicalization are added to the TODO list.

**Comment by varkor** [comment]
The refactoring is a separate issue: it's not a blocker for const generics in any way, though we'll certainly have to revisit canonicalisation.

**Comment by varkor** [comment]
With https://github.com/rust-lang/rust/pull/59008 merged, all of the changes from this pull request have been integrated into the main codebase. However, there's some slight sentiment towards this pull request, having been open so long, so we plan to use it to address a few final issues as a token of the first major milestone towards const generics. I'll update the pull request soon.

**Comment by mark-i-m** [comment]
@varkor Regarding tests, perhaps it would help to know what is expected to work. For example, I was surprised that this does not work: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=d84ffd15226fcffe02c102edb8ae5cf1

**Comment by varkor** [comment]
I've resolved several remaining `FIXME`s. Let's leave it there and resolve the remaining issues in separate pull requests.

r? @eddyb

File: src/librustc/infer/canonical/query_response.rs
Comment (eddyb) [inline src/librustc/infer/canonical/query_response.rs]: No, this needs to be an  `eq` just like the type one above.

File: src/librustc/infer/canonical/query_response.rs
Comment (eddyb) [inline src/librustc/infer/canonical/query_response.rs]: Same here, this needs to be an `eq`.

File: src/librustc_mir/monomorphize/item.rs
Comment (eddyb) [inline src/librustc_mir/monomorphize/item.rs]: This whole infrastructure should be rewritten to use `ty::print`, can you open an issue about that?

**Comment by mark-i-m** [comment]
> _But as they gazed upon that which claimed to be the Beginning,_

_it was rebased, the inevitable conclusion of all epic journeys in the land of git. And thus concluded the Genesis of Generic Germination_.

File: src/librustc_mir/monomorphize/item.rs
Comment (varkor) [reply src/librustc_mir/monomorphize/item.rs]: Opened https://github.com/rust-lang/rust/issues/60582.

**Review by eddyb — APPROVED** [review]

**Pull request merged**
